DOCUMENTATION.md

```diff
@@ -0,0 +1,110 @@
+
+In the context of ZK-proofs, choosing $d$ is an important task, given that it will affect the efficiency of the proof generation. The algorithms considered in the library require approximately $O(N^{1.5})$ multiplications and additions and $O(l)$ range checks. As we mentioned before, $d$ is inversely proportional to $N$. Hence, a large $d$ would be a good choice for a large modulus. However, if the modulus is small, it is best to choose a small $d$ to reduce the complexity of the algorithms.
+
+In the rest of the document and the implementation, we will consider $d = 120$ as a reasonable choice of parameters. Also, in the implementation, we consider $N$ to be fixed because of the technical limitations of Noir, as the library does not support an efficient mechanism to store the limbs dynamically. Although the library support Rust-like vectors, they are not very efficient and pur goal is also to implement something that can be used in production. Hence, we are limited to using fixed-length arrays to represent the collection of limbs for a number. To give more freedom during circuit development, the API is designed so that the programmer can set the value of $N$ at compile time.
+
+## Big-integer representation
+
+The building block to construct a big-integer library is the BigNum struct:
+
+```rust
+struct BigNum<N,Params> {
+    limbs: [Field; N],
+}
+```
+
+This struct represents a big-integer with $N$ limbs. Each limb will be represented as a `Field` element whose value is in the range $\{0, \dots, 2^d - 1\}$ (remember that in this case, $d = 120$). We also say that the number is represented in a radix of $d$ bits. Considering this representation, if a number is represented using the vector $(a_0, \dots, a_{N-1})$, then, the decimal representation of this number will be $\sum_{i=0}^{N-1} a_i \cdot 2^{d \cdot i}$.
+
+The API is designed around this struct so the addition and multiplication are defined as methods implemented for this struct.
+
+## Utilities
+
+During the library's implementation, we require some utilities that allow us to implement the arithmetic operations in an idiomatic way. Here, we describe those utilities and their usage inside the main arithmetic algorithms.
+
+### `ArrayX`
+
+The struct `ArrayX` is an implementation of an array whose length is the product of a known multiplier `SizeMultiplier` and $N$. For example, if we define the value of $N$ at compile time and need to store twice that amount of values, we create an `ArrayX` where `SizeMultiplier` is two. This implementation is a workaround to have arrays whose length is $c \cdot N$ for a known constant $c$ because, in our implementation, the value of $N$ is considered a generic, and it is not possible to operate over generics to create an array of size $c \cdot N$.
+
+The struct for `ArrayX` is presented next:
+
+```rust
+struct ArrayX<T, N, SizeMultiplier> {
```

Comment:
When we switch to a new Noir version, these snippets will be slightly different because of the numeric generics

Reply:
Changed to
```rust
struct ArrayX<T, let N: u64, let SizeMultiplier: u8> {
    segments: [[T; N]; SizeMultiplier]
}
```